### PR TITLE
Buffer Safe String Functions

### PR DESCRIPTION
**Pull request by trevorheaws**
Replacing string functions with buffer checked and automatically null terminated versions.

### CODE REVIEW

File: prov/bgq/src/fi_bgq_pmi.c [hunk 1]
```diff
@@ -338,24 +341,29 @@ int PMI_KVS_Get (const char kvsname[], const char key[], char value[], int lengt
...
 				current_block_id += current_node_block_count;
-				node_block_write(value, current_node_block_value, current_block_id, 1, current_proc_count);
+				node_block_write(value, length, current_node_block_value, sizeof(current_node_block_value),
+						current_block_id, 1, current_proc_count);
 			}
 		}
 		// add closing bracket to value
-		strcat(value,")");
+		int offset = strlen(value);
+		snprintf(value+offset, length-offset, ")");
```
Comment (shefty) [inline prov/bgq/src/fi_bgq_pmi.c:366]: I would drop all changes to the bgq provider.  All development on this provider has been halted.

**Comment by trevorheaws** [comment]
@shefty Thank you for reviewing this request.  Would it be possible to get some details about the failing continuous-integration/jenkins/pr-head the http://sfs-login.jf.intel.com/ link provided does not work for me.

**Comment by shefty** [comment]
```
prov/psm2/src/psmx2_init.c: In function 'psmx2_update_hfi_info':
16:55:58  prov/psm2/src/psmx2_init.c:395:47: warning: left-hand operand of comma expression has no effect [-Wunused-value]
16:55:58     snprintf((psmx2_hfi_info.default_domain_name, sizeof(psmx2_hfi_info.default_domain_name) - offset,
16:55:58                                                 ^
16:55:58  prov/psm2/src/psmx2_init.c:395:100: warning: left-hand operand of comma expression has no effect [-Wunused-value]
16:55:58     snprintf((psmx2_hfi_info.default_domain_name, sizeof(psmx2_hfi_info.default_domain_name) - offset,
16:55:58                                                                                                      ^
16:55:58  prov/psm2/src/psmx2_init.c:396:8: warning: left-hand operand of comma expression has no effect [-Wunused-value]
16:55:58      "%s", unit_name);
16:55:58          ^
16:55:58  prov/psm2/src/psmx2_init.c:396:20: error: expected ')' before ';' token
16:55:58      "%s", unit_name);
16:55:58                      ^
16:55:58  prov/psm2/src/psmx2_init.c:400:2: error: too few arguments to function 'snprintf'
16:55:58    }
16:55:58    ^
16:55:58  prov/psm2/src/psmx2_init.c:400:2: error: expected ';' before '}' token
16:55:58  prov/psm2/src/psmx2_init.c:284:6: warning: variable 'multirail' set but not used [-Wunused-but-set-variable]
16:55:58    int multirail = 0;
16:55:58        ^
16:55:58  make[1]: *** [prov/psm2/src/libpsmx2_la-psmx2_init.lo] Error 1
16:55:58  make[1]: Leaving directory `/jenkinsbuild/workspace/ofi_libfabric_PR-6226'
```

There's a double (( after snprintf.

File: prov/psm2/src/psmx2_init.c [hunk 2]
```diff
@@ -387,10 +387,13 @@ static int psmx2_update_hfi_info(void)
...
-		sprintf(unit_name, "hfi1_%hu", i);
+		snprintf(unit_name, sizeof(unit_name), "hfi1_%hu", i);
+		int offset = 0;
 		if (psmx2_hfi_info.num_active_units > 1)
-			strcat(psmx2_hfi_info.default_domain_name, ";");
-		strcat(psmx2_hfi_info.default_domain_name, unit_name);
+			offset = snprintf(psmx2_hfi_info.default_domain_name,
+					sizeof(psmx2_hfi_info.default_domain_name), ";");
+		snprintf((psmx2_hfi_info.default_domain_name, sizeof(psmx2_hfi_info.default_domain_name) - offset,
+			"%s", unit_name);
```
[the remark this answers is not in the document]
Comment (trevorheaws) [reply prov/psm2/src/psmx2_init.c:397]: fixed

File: prov/util/src/util_attr.c [hunk 3]
```diff
@@ -95,12 +95,13 @@ int ofi_exclude_prov_name(char **prov_name_list, const char *util_prov_name)
...
 
-	exclude = malloc(strlen(util_prov_name) + 2);
+	int length = strlen(util_prov_name) + 2;
+	exclude = malloc(length);
 	if (!exclude)
 		return -FI_ENOMEM;
 
 	exclude[0] = '^';
-	strcpy(&exclude[1], util_prov_name);
+	snprintf(&exclude[1], length - 1, util_prov_name); // The -1 accounts for the '^' character.
```
[the remark this answers is not in the document]
Comment (trevorheaws) [reply prov/util/src/util_attr.c]: Reworked so it is a bit simpler and under 80 character line length.

**Comment by rwespetal** [comment]
ok to test

**Comment by rwespetal** [comment]
add to whitelist

File: prov/util/src/util_attr.c [hunk 4]
```diff
@@ -95,12 +95,13 @@ int ofi_exclude_prov_name(char **prov_name_list, const char *util_prov_name)
 {
 	char *exclude, *name, *temp;
 
-	exclude = malloc(strlen(util_prov_name) + 2);
+	int length = strlen(util_prov_name) + 2;
```
[the remark this answers is not in the document]
Comment (trevorheaws) [reply prov/util/src/util_attr.c]: fixed

File: src/unix/osd.c [hunk 5]
```diff
@@ -116,12 +116,12 @@ int ofi_shm_map(struct util_shm *shm, const char *name, size_t size,
 	*mapped = MAP_FAILED;
 	memset(shm, 0, sizeof(*shm));
 
-	fname = calloc(1, strlen(name) + 2); /* '/' + %s + trailing 0 */
+	int fname_size = strlen(name) + 2; /* '/' + %s + trailing 0 */
```
[the remark this answers is not in the document]
Comment (trevorheaws) [reply src/unix/osd.c]: fixed

File: prov/bgq/src/fi_bgq_pmi.c (hunk 1 above)
Comment (trevorheaws) [reply prov/bgq/src/fi_bgq_pmi.c]: I pulled out this commit.

**Comment by trevorheaws** [comment]
@shefty I am still seeing failures on the intel build, presumably with a different error. Would you be able to provide the log again. Thank you!

File: util/pingpong.c [hunk 6]
```diff
@@ -1869,7 +1869,7 @@ static int pp_finalize(struct ct_pingpong *ct)
 
 	PP_DEBUG("Terminating test\n");
 
-	strcpy(ct->tx_buf, fin_buf);
+	snprintf(ct->tx_buf, fin_buf_size, fin_buf);
```
[the remark this answers is not in the document]
Comment (trevorheaws) [reply util/pingpong.c:1872]: Done

**Comment by shefty** [comment]
The Intel build failure was some issue with MPICH.

**Pull request merged**